### ISSUE TITLE
the loss function is not useful in the experiment?

### ISSUE THREAD

**Issue by xiajun112233** [bboylyg/NAD#4]
Hello, I'm very interested in this paper, when I see the main.py,  the three at_loss are used .detach() to out of the calculate graph in the PyTorch. So I delete the at1_loss、at2_loss、at3_loss in the loss function. But, when I run the changed code, the ASP is still very low.  I think the at_loss is not useful in the code. The training dataset in the main code is the clean dataset, not the backdoor dataset, so the NAD ASR is very low. However, the training dataset in the train_badnets code uses the backdoor dataset, so the baseline ASR is high. I changed the training dataset in the main code to the backdoor dataset. Unfortunately, the NAD is not useful in the backdoor dataset.
![image](https://user-images.githubusercontent.com/24726126/124412677-9ed1c780-dd81-11eb-8a2e-a895eb5ee83c.png)

**Comment by bboylyg** [comment]
Hi, Thanks for your interest in our work. To verify the effectiveness of NAD, you could finetune the backdoored student with/without the NAD loss, i.e. setting at1_loss, at2_loss, and at3_loss all to be non-zero/zero, and compare the ASR under two types of settings.

**Comment by xiajun112233** [comment]
without at_loss
![image](https://user-images.githubusercontent.com/24726126/124417099-017b9100-dd8b-11eb-801a-13a4c29d7c81.png)
![image](https://user-images.githubusercontent.com/24726126/124417196-2f60d580-dd8b-11eb-9076-e1299162f0b4.png)

with at_loss
![image](https://user-images.githubusercontent.com/24726126/124417240-46072c80-dd8b-11eb-9820-f01863fe7309.png)
![image](https://user-images.githubusercontent.com/24726126/124419710-a351ac80-dd90-11eb-91d5-b31d6c8da74e.png)

**Comment by bboylyg** [comment]
Thanks for providing the screenshot. It is clear to see that there achieves a better erasing result with NAD loss(ASR decreases to 3.78\%, compared to the result without NAD loss).  By the way, the selection of `trigger types\teacher models\data augmentation techniques` also causes different erasing effects for distillation.

**Comment by xiajun112233** [comment]
But, when I run without NAD loss train code, there also have good results in ASR, so I think it is random results for the CE loss in the clean dataset, you can see the next pictures.  Whether use the clean dataset to retrain the backdoor model is good enough to defend against the backdoor attack? Thank you.
![image](https://user-images.githubusercontent.com/24726126/124435415-8bd1ee00-dda7-11eb-8c53-047411eefd59.png)

![image](https://user-images.githubusercontent.com/24726126/124435099-35fd4600-dda7-11eb-9c36-17a2c33f39bc.png)

**Comment by bboylyg** [comment]
To be honest, It is not surprising that Fine-tuning can effectively erase BadNets attack; the erasing effect is probably attributed to the data augmentation techniques, i.e. Padding, flip, and cutout, as they are highly related to the original trigger pattern. You can change the param of `Cutout` as 1 hole with a litter size 9 or 4 to verify this observation. By the way, I think the `adaptive attacks` shown in `Appendix K(Table 9)` in our paper will be beneficial to your understanding of our NAD.

**Comment by xiajun112233** [comment]
OK, thank you, which parameters in the code should I change to use the adaptive attacks in this code?

**Comment by bboylyg** [comment]
The most simple case is that changing the location of the backdoor trigger (i.e. BadNets trigger) from the bottom-right to the center of the image.
